apps/backend/src/mutations/FapMutations.ts

```diff
@@ -243,62 +244,108 @@ export default class FapMutations {
       );
 
     const callIds = [...new Set(proposals.map((proposal) => proposal.callId))];
     for (const callId of callIds) {
-      const callHasInstrument = callHasInstruments.find(
-        (callHasInstrument) => callHasInstrument.callId === callId
-      );
-      if (callHasInstrument && callHasInstrument.fapId) {
-        await this.assignProposalsToFapInternal(agent, {
-          proposals: proposals
-            .filter((proposal) => proposal.callId === callId)
-            .map((proposal) => ({
-              ...proposal,
-              callId: callId,
-            })),
-          fapId: callHasInstrument.fapId,
-          fapInstrumentId: callHasInstrument.instrumentId,
-        });
+      if (callHasInstruments.length) {
+        const fapInstruments = callHasInstruments
+          .filter((callHasInstrument) => callHasInstrument.fapId)
+          .map((callHasInstrument) => ({
+            fapId: callHasInstrument.fapId,
+            instrumentId: callHasInstrument.instrumentId,
+          }));
+        if (fapInstruments.length) {
+          await this.assignProposalsToFapsInternal(agent, {
+            proposalPks: proposals
+              .filter((proposal) => proposal.callId === callId)
+              .map((proposal) => proposal.primaryKey),
+            fapInstruments: fapInstruments,
+          });
+        }
       }
     }
 
     return true;
   }
 
   @Authorized([Roles.USER_OFFICER])
-  async assignProposalsToFap(
+  async assignProposalsToFaps(
     agent: UserWithRole | null,
-    args: AssignProposalsToFapArgs
+    args: AssignProposalsToFapsArgs
   ): Promise<ProposalPks | Rejection> {
-    return this.assignProposalsToFapInternal(agent, args);
+    return this.assignProposalsToFapsInternal(agent, args);
   }
 
-  @EventBus(Event.PROPOSAL_FAP_SELECTED)
-  async assignProposalsToFapInternal(
+  @EventBus(Event.PROPOSAL_FAPS_SELECTED)
+  async assignProposalsToFapsInternal(
     agent: UserWithRole | null,
-    args: AssignProposalsToFapArgs
+    args: AssignProposalsToFapsArgs
   ): Promise<ProposalPks | Rejection> {
-    const result = await this.dataSource.assignProposalsToFap(args);
+    if (!args.fapInstruments.length) {
+      return rejection(
+        'Proposal cannot be assigned to FAP without specifying the instrument and FAP',
+        {
+          agent,
+        }
+      );
+    }
+
+    const dataToInsert: AssignProposalsToFapsInput[] = [];
+
+    for (const proposalPk of args.proposalPks) {
+      for (const fapInstrument of args.fapInstruments) {
+        const proposalAssignedInstruments =
+          await this.instrumentDataSource.getInstrumentsByProposalPk(
+            proposalPk
+          );
 
-    if (result.proposalPks.length !== args.proposals.length) {
+        const fullProposal = await this.proposalDataSource.get(proposalPk);
+
+        if (!fullProposal) {
+          return rejection(`Proposal not found with id: ${proposalPk}`, {
+            args,
+          });
+        }
```

Comment:
Can this code be moved out of the innermost loop? It looks like it's only dependent on the variable of the outer one.